### DISCUSSION TITLE
Use median instead of mean for last play offset adjustment

### DISCUSSION THREAD

**Discussion by tleydxdy** [ppy/osu#19763] [Ideas]
Using median should reduce the effect of player error on the result and get a truer offset value. It also intuitively make sense to adjust the offset where half of the hits are early and half are late.

**Comment by peppy** [comment]
Can agree with this.

**Reply by peppy** [reply]
Thinking about it more I'm actually not sure. It may end up being worse for cases where beatmaps have multiple timing points and some are mistimed more than others. Will wait for more opinions/discussion on the matter.

If anyone wishes to test this, you can use this branch https://github.com/ppy/osu/compare/master...peppy:osu:median-error?expand=1

or patch:

```diff
diff --git a/osu.Game/Rulesets/Scoring/HitEventExtensions.cs b/osu.Game/Rulesets/Scoring/HitEventExtensions.cs
index 2fde73d5a2..2d2562eeae 100644
--- a/osu.Game/Rulesets/Scoring/HitEventExtensions.cs
+++ b/osu.Game/Rulesets/Scoring/HitEventExtensions.cs
@@ -24,6 +24,27 @@ public static class HitEventExtensions
             return 10 * standardDeviation(timeOffsets);
         }
 
+        /// <summary>
+        /// Calculates the median hit offset/error for a sequence of <see cref="HitEvent"/>s, where negative numbers mean the user hit too early on average.
+        /// </summary>
+        /// <returns>
+        /// A non-null <see langword="double"/> value if unstable rate could be calculated,
+        /// and <see langword="null"/> if unstable rate cannot be calculated due to <paramref name="hitEvents"/> being empty.
+        /// </returns>
+        public static double? CalculateMedianHitError(this IEnumerable<HitEvent> hitEvents)
+        {
+            double[] timeOffsets = hitEvents
+                                   .Where(affectsUnstableRate)
+                                   .Select(ev => ev.TimeOffset)
+                                   .OrderBy(x => x)
+                                   .ToArray();
+
+            if (timeOffsets.Length == 0)
+                return null;
+
+            return timeOffsets[timeOffsets.Length / 2];
+        }
+
         /// <summary>
         /// Calculates the average hit offset/error for a sequence of <see cref="HitEvent"/>s, where negative numbers mean the user hit too early on average.
         /// </summary>
diff --git a/osu.Game/Screens/Play/PlayerSettings/BeatmapOffsetControl.cs b/osu.Game/Screens/Play/PlayerSettings/BeatmapOffsetControl.cs
index 75da8e7b9d..dd37175249 100644
--- a/osu.Game/Screens/Play/PlayerSettings/BeatmapOffsetControl.cs
+++ b/osu.Game/Screens/Play/PlayerSettings/BeatmapOffsetControl.cs
@@ -190,7 +190,7 @@ private void scoreChanged(ValueChangedEvent<ScoreInfo> score)
 
             var hitEvents = score.NewValue.HitEvents;
 
-            if (!(hitEvents.CalculateAverageHitError() is double average))
+            if (!(hitEvents.CalculateMedianHitError() is double average))
                 return;
 
             referenceScoreContainer.Children = new Drawable[]

```

Note that the formula is roughly incorrect for event result coutns but should be enough to test whether it works well or not.

**Reply by tleydxdy** [reply]
ah, sorry, I wasn't thinking about the automatic adjustment, I meant that button you can press when you retry a map. tho it could make sense for auto adjustment as well, but that should be changed more carefully for sure.

**Reply by peppy** [reply]
I'm not sure I follow. i'm pretty sure you're talking about the same control.

**Reply by tleydxdy** [reply]
you're right, I should have expanded the diff and read further.

**Reply by sineplusx** [reply]
Is this still something that the dev team would want to see implemented? I did some testing and found it works well to reduce the offset skew by player error to some degree (in comparison to just using the mean).

One downside is that the median needs a large sample size to work well, at a low number of `HitEvents` the offset gets very inaccurate. Right now the offset calibration needs 10 timed hits to become available, that number would probably need to be raised.

**Reply by peppy** [reply]
I still fail to understand how, so please ELI5 using an example case before we even consider this.

**Reply by sineplusx** [reply]
The offset calibration can only work well if you actually time hits to the beat of the music. With difficult maps that are outside of a players comfort zone, they might hit disproportionally early or late. These outlier values can skew the average value in one direction, resulting in a shifted offset if applied. Using the median (central value) instead of the mean can reduce that skew by outliers.

Here is an example where I botched a stream section in a map, followed by a normal playthrough. Both were played without any offset preapplied.

| Bad Run | Normal Run |
|--------|--------|
| ![bad-run](https://github.com/user-attachments/assets/5ea0a5ef-9c54-4e60-a504-4855d0ab9fb6) | ![good-run](https://github.com/user-attachments/assets/9ef22fa4-ea1b-4a0c-8e0b-e015975df54a) |

![boxplot](https://github.com/user-attachments/assets/9819ea27-4454-4765-97e3-e3c34903dfa0)

|  | n | σ | Mean | Median | 
|--------|--------|--------|--------|--------|
| Bad Run (Top) | 1194 | 14.1219 | -4.8298 | -3.3829 |
| Normal Run (Bottom) | 1204 | 9.3441 | -2.8382 | -2.6325 |

In the "Bad run" scenario, using the offset calibration after the run would currently result in an error of  ~2.2 ms to the true offset.

**Reply by peppy** [reply]
Thanks for the excellent explanation.

I think you could go ahead with attempting this change, it sounds worthwhile.

**Reply by sineplusx** [reply]
#32601
